### ISSUE TITLE
strictNullCheck should detect unreachable default values and useless nullish operator

### ISSUE THREAD

**Issue by marcandre** [microsoft/TypeScript#43040]
# Suggestion

Using `strictNullCheck` should allow detection of code that isn't supposed to be reached according to type analysis

## 🔍 Search Terms

strictNullCheck unreachble nullish default value

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

Yes, especially "Statically identify constructs that are likely to be errors."

## ⭐ Suggestion

This assumes that it is desired to detect errors with unreachable code due to type analysis (see #43038).

With `strictNullCheck` turned on, some code that is meant to be unreachable and thus an error is not flagged.

## 📃 Motivating Example

The `= 42` below is unreachable ([playground](https://www.typescriptlang.org/play?#code/MYGwhgzhAECiAeYC2AHEBTaBvAUNaAZgPZEAUu++80AvNACwBMe0AvgFzYtUD8nAdgFckAI3QAnFqwCUXStHHoALoPH9o8KTlY4gA)) and is the sign that some code is likely to be an error, but nothing is flagged

```typescript
class Example {
  foo({
    x = 42
  }: {
    x: number
  }) {
    return x
  }
}
```

Either the declaration should be changed to `x?: number`, or the default should be removed.

Another example ([thanks to @jcalz](https://stackoverflow.com/questions/66432817/why-isnt-it-an-error-to-provide-a-default-value-that-cant-be-used)) using the nullish operator.

```typescript
function foo(x: number) { x ??= 42 }
```

Again, it would be helpful if an error or a warning would be raised as that code seems clearly to be a mistake.

**Comment by MrAndersen1** [comment]
Just wanted to say your playground link has x?: number in it instead of the expected x: number

**Comment by marcandre** [comment]
@MrAndersen1 thanks, updated.

**Comment by zoontek** [comment]
I can agree more on this feature. Currently, this is valid TS code:

```ts
let foo = "foo";
let bar = foo ?? 42;
```

`bar` will never be of type `string | number`, the compiler should return an error since `foo` is not of type `string | null`, `string | undefined` or `string | null | undefined`
